caikit_nlp/toolkit/data_stream_wrapper.py

```diff
@@ -48,10 +50,26 @@ def __init__(self, stream, shuffle, buffer_size=None):
         log.debug("Shuffling buffer size: {}".format(self.buffer_size))
 
     def __iter__(self):
-        if self.shuffle:
-            log.debug4("Reshuffling training data!")
-            return iter(self.stream.shuffle(self.buffer_size))
-        return iter(self.stream)
+        worker_info = get_worker_info()
+
+        if worker_info is None:  # single-process data loading, return the full iterator
+            if self.shuffle:
+                log.debug4("Reshuffling training data!")
+                return iter(self.stream.shuffle(self.buffer_size))
+            return iter(self.stream)
+
+        # When num_workers > 0, each worker process will have a different copy of the dataset object,
+        # so we configure each copy independently to avoid having duplicate data returned from
+        # each worker
+        else:  # in a worker process
+            # split workload
+            per_worker = int(
+                math.ceil((self.end - self.start) / float(worker_info.num_workers))
+            )
+            worker_id = worker_info.id
+            iter_start = self.start + worker_id * per_worker
+            iter_end = min(iter_start + per_worker, self.end)
+        return iter(range(iter_start, iter_end))
```

Comment:
you mean add for multi-device case? that part is tricky since we won't have multi-device available in testing environment. I guess, I can patch the `get_worker_info` from transformers to write the test for this 🤔